docs/tutorials/05_kv-scheduler.md

```diff
@@ -0,0 +1,303 @@
+ 
+ #### 1. Adapters
+ 
+ Let's start with adapters. The purpose of the adapter is to define conversion methods between our proto-defined type and bare `proto.Message`. Since this code fulfils the definition of a boilerplate, we will generate it. The generator is called `descriptor-adapter` and can be found in the [inside the KVScheduler plugin][2]. Build the binary file from the go files inside, and use it to generate two adapters for the `Interface` and `Route` proto messages:
+ 
+```
+descriptor-adapter --descriptor-name Interface --value-type *model.Interface --import "github.com/ligato/vpp-agent/examples/tutorials/05_kv-scheduler/model" --output-dir "descriptor"
+descriptor-adapter --descriptor-name Route --value-type *model.Route --import "github.com/ligato/vpp-agent/examples/tutorials/05_kv-scheduler/model" --output-dir "descriptor"
+```
+
+It is a good practice to add those commands to the plugin file with the `//go:generate` directives. Adapters were automatically created in the `adapter` directory within plugin folder.
+
+#### 2. Descriptor without dependency
+
+Another step is to define descriptors. We start with the interface descriptor, since there is no dependency on other messages. Let's create the `descriptors.go` so the code is outside of `main.go`, and define following code:
+
+```go
+type IfDescriptor struct {
+	log logging.PluginLogger
+}
+
+func NewIfDescriptor(logger logging.PluginLogger) *IfDescriptor {
+	return &IfDescriptor{
+		log: logger,
+	}
+}
+```
+
+The `IfDescriptor` is a descriptor object which can define its own dependencies (used in descriptor methods). The `NewIfDescriptor` is a constructor (and also initializes descriptor dependencies). Next we define `GetDescriptor()` method which returns typed descriptor (from generated adapter code):
+
+```go
+func (d *IfDescriptor) GetDescriptor() *adapter.InterfaceDescriptor {
+    return &adapter.InterfaceDescriptor{}
+}
+```
+
+If you have a look at `adapter.InterfaceDescriptor`, you see that it defines several fields. Most important fields are function-types with CRUD definition and fields resolving dependencies. The full API list is documented in the [KvDescriptor structure][3]. Here, we implement the necessary ones:
+
+* Name of the descriptor, must be unique for all descriptors. 
+```go
+    Name: "if-descriptor",
+```
+
+* Northbound key prefix for configuration type handled by the descriptor
+```go
+NBKeyPrefix: "/interface/",
+```
+
+* String representation of the type
+```go
+ValueTypeName: proto.MessageName(&model.Interface{}),
+```
+
+* Configuration item identifier (label, name, index) is returned by this method. 
+```go
+KeyLabel: func(key string) string {
+    return strings.TrimPrefix(key, "/interface/")
+},
+```
+
+* Key selector returns true if the provided key is valid for given descriptor. In the example, every descriptor uses own key, but there are cases where the descriptor processes other descriptor's keys or derived types.
+```go
```

Comment:
This is quite confusing - "but there are cases where the descriptor processes other descriptor's keys or derived types". The selectors should ensure that each descriptor selects a distinct subset of keys. Furthermore there is a 1:1 mapping between value-type and descriptor (i.e. no more than one value type processed by any descriptor).
Also, the method below should return `return strings.TrimPrefix(key, "/interface/")` and not just `true`.

Reply:
rephrased, method set with the value from the example